daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala

```diff
@@ -261,6 +262,10 @@ abstract class SequenceParserBase(
               // So we mask the failure, and exit the sequence successfully
               pstate.setSuccess()
               isDone = true
+              // If we're masking the failure, we don't want the error dianostics
+              // to flow up. Restore the diagnostics from before the parse
+              // attempt
+              pstate.diagnostics = diagnosticsBeforeAttempt
             }
 
             // We successfully parsed a discriminator, but failed to parse the discriminated content.
```

Comment:
I noticed that this match expression has 4 cases and I looked at each of them.  The first 3 cases set isDone = true, but we've added pstate.diagnostics = diagnosticsBeforeAttempt to only the 1st and 3rd cases.  Should the 2nd case (which begins on line 271) also have pstate.diagnostics = diagnosticsBeforeAttempt added after isDone = true?

Reply:
I think the only place it's needed is when we do ``pstate.setSuccess()``. That's where we are essentially ignoring any previous errors, and so we also need to remove diagnostics that came from those errors.

I think part of the fix to DAFFODIL-2512 might want to change how this works. Something like rather than overriding failed state and diagnostics, maybe we want a PoU and we just reset back to it. But that might undo other state, like infoset stuff, that we do want to keep around. So it's not so obvious what the right thing to do is here.